include/zephyr/posix/sys/dirent.h

```
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2024 Tenstorrent AI ULC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_POSIX_SYS_DIRENT_H_
+#define ZEPHYR_INCLUDE_POSIX_SYS_DIRENT_H_
+
+#include <limits.h>
+
+#include <zephyr/posix/posix_features.h>
+
+#ifndef PATH_MAX
+#ifdef _XOPEN_SOURCE
+#define PATH_MAX _XOPEN_PATH_MAX
+#endif
+#endif
+
+#ifndef PATH_MAX
+#ifdef _POSIX_C_SOURCE
+#define PATH_MAX _POSIX_PATH_MAX
+#endif
+#endif
+
+#ifndef PATH_MAX
+#define PATH_MAX 31
```

We have FILE_SYSTEM_MAX_FILE_NAME https://github.com/search?q=repo%3Azephyrproject-rtos%2Fzephyr%20FILE_SYSTEM_MAX_FILE_NAME&type=code, and that will impact the PATH_MAX if set above 31 characters.

---

Now that I'm looking at this, it really seems like this should be `NAME_MAX` instead of `PATH_MAX`, which has corresponding `_XOPEN_NAME_MAX` and `_POSIX_NAME_MAX` [ref](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/limits.h.html)

Indeed, that seems to be true [ref](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/dirent.h.html).

Interestingly, this relates to `_POSIX_NO_TRUNC`. If that value is `> 0` and a path component (name) longer than `_NAME_MAX` is supplied to a function, it will return with an error code `ENAMETOOLONG` [ref](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/unistd.h.html). Otherwise, the path component is simply truncated. Currently, `_POSIX_NO_TRUNC` is zero, so we're OK to silently truncate.

> We have FILE_SYSTEM_MAX_FILE_NAME

It looks as though `FILE_SYSTEM_MAX_FILE_NAME` appllies to the complete path instead of individual path components (names).

I did a quick search [here](https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_FILE_SYSTEM_MAX_FILE_NAME) with the search `^CONFIG_FILE_SYSTEM_.*$` and it looks like there isn't anything corresponding to `NAME_MAX` within Zephyr's native FS subsystem.

---

Fixed! (with less code, as it turns out)

---

> I did a quick search [here](https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_FILE_SYSTEM_MAX_FILE_NAME) with the search ^CONFIG_FILE_SYSTEM_.*$ and it looks like there isn't anything corresponding to NAME_MAX within Zephyr's native FS subsystem.

We actually do not have anything corresponding to PATH_MAX, there are app specific defines like in fs shell, but FS system itself does not have such. The `CONFIG_FILE_SYSTEM_MAX_FILE_NAME ` is only used by `fs_dirent` to return name of file found while traversing directory contents - that is the single place.
We do not have anything like `PATH_MAX` due to how the FS processes paths: after stripping the mount point we are giving full path, that user has created/provided, to file system driver and the driver then goes through the path and tries to figure out where the file is, so from the point of view of Zephyr the path can be infinitely long and we do not care.
Zephyr VFS does not store paths anywhere, they are left on the user/application side.

> Interestingly, this relates to _POSIX_NO_TRUNC. If that value is > 0 and a path component (name) longer than _NAME_MAX is supplied to a function, it will return with an error code ENAMETOOLONG [ref](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/unistd.h.html). Otherwise, the path component is simply truncated. Currently, _POSIX_NO_TRUNC is zero, so we're OK to silently truncate.

Not sure this is fine, since from time to time we may be actually lucky to open/remove/truncate some unintended file.